### PR TITLE
MART typing

### PR DESCRIPTION
**Pull request by lumurillo**
# What does this PR do?

Add typing annotation to the MART implementation to make the code more readable and robust.

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] `pytest --cov=mart`

## Before submitting

- [x] The title is **self-explanatory** and the description **concisely** explains the PR
- [x] My **PR does only one thing**, instead of bundling different changes together
- [ ] I list all the **breaking changes** introduced by this pull request
- [x] I have commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have run pre-commit hooks with `pre-commit run -a` command without errors

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: mart/attack/projector.py [hunk 1]
```diff
@@ -103,20 +131,30 @@ class LinfAdditiveRange(Projector):
     """Make sure the perturbation is within the Linf norm ball, and "input + perturbation" is
     within the [min, max] range."""
 
-    def __init__(self, eps, min=0, max=255):
+    def __init__(self, eps: float, min: Optional[int] = 0, max: Optional[int] = 255):
```
Comment (dxoigmn) [inline mart/attack/projector.py:134]: In theory `min` and `max` can be `float`s.

File: mart/attack/adversary.py [hunk 2]
```diff
@@ -19,29 +25,71 @@ class AdversaryCallbackHookMixin(Callback):
 
     callbacks = {}
 
-    def on_run_start(self, adversary, input, target, model, **kwargs):
+    def on_run_start(
+        self,
+        adversary: Callback,
```
Comment (dxoigmn) [inline mart/attack/adversary.py:30]: I think we would want to type this as `Adversary`?

File: mart/attack/adversary.py [hunk 3]
```diff
@@ -19,29 +25,71 @@ class AdversaryCallbackHookMixin(Callback):
...
+        input: Union[torch.Tensor, tuple],
+        target: Union[torch.Tensor, Dict[str, Any], tuple],
+        model: SequentialDict,
```
Comment (dxoigmn) [inline mart/attack/adversary.py:33]: I think this should be more generally typed as a `torch.nn.Module`.

Comment (dxoigmn) [reply mart/attack/adversary.py:33]: At least, I don't think we make assumptions about `model` being a `SequentialDict`.

File: mart/attack/adversary_in_art.py [hunk 4]
```diff
@@ -21,7 +24,7 @@ class MartToArtAttackAdapter:
     but it should be reusable for other models with slight modifications.
     """
 
-    def __init__(self, target_model, mart_exp_config_yaml, **kwargs):
+    def __init__(self, target_model: torch.nn.Module, mart_exp_config_yaml: str, **kwargs):
```
Comment (dxoigmn) [inline mart/attack/adversary_in_art.py:27]: I think this class should adopt the types in https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/987052c405e05d458276299aafc7d47bb584e738/art/estimators/object_detection/object_detector.py#L42

File: mart/attack/adversary_wrapper.py [hunk 5]
```diff
@@ -16,7 +20,7 @@ class NormalizedAdversaryAdapter(torch.nn.Module):
     External adversaries commonly take input of NCWH-[0,1] and return input_adv in the same format.
     """
 
-    def __init__(self, adversary):
+    def __init__(self, adversary: Callback):
```
Comment (dxoigmn) [inline mart/attack/adversary_wrapper.py:23]: I don't think this is the correct type. This should really be named `external_adversary` so as not to be confused with MART's implementation of `Adversary`. The only thing we assume about `external_adversary` is that it is `Callable`.

File: mart/attack/callbacks/base.py [hunk 6]
```diff
@@ -6,27 +6,74 @@
...
 class Callback(abc.ABC):
     """Abstract base class of callbacks."""
 
-    def on_run_start(self, adversary, input, target, model, **kwargs):
+    def on_run_start(
+        self,
+        adversary: torch.nn.Module,
+        input: Union[torch.Tensor, tuple],
+        target: Union[torch.Tensor, Dict[str, Any], tuple],
+        model: SequentialDict,
```
Comment (dxoigmn) [inline mart/attack/callbacks/base.py:26]: I think this should be `nn.Module` too?

File: mart/attack/callbacks/base.py [hunk 7]
```diff
@@ -6,27 +6,74 @@
...
 __all__ = ["Callback"]
 
 
 class Callback(abc.ABC):
     """Abstract base class of callbacks."""
 
-    def on_run_start(self, adversary, input, target, model, **kwargs):
+    def on_run_start(
+        self,
+        adversary: torch.nn.Module,
```
Comment (dxoigmn) [inline mart/attack/callbacks/base.py:23]: This should be `Adversary`.

File: mart/attack/gain.py [hunk 8]
```diff
@@ -20,14 +22,14 @@ class RoIHeadTargetClass(Gain):
...
-    def __init__(self, class_index=0, targeted=True) -> None:
+    def __init__(self, class_index: Optional[int] = 0, targeted: Optional[bool] = True) -> None:
         super().__init__()
 
         self.gain = torch.nn.CrossEntropyLoss()
         self.class_index = class_index
         self.targeted = targeted
 
-    def forward(self, roi_heads_class_logits, proposals):
+    def forward(self, roi_heads_class_logits: torch.Tensor, proposals: Any) -> torch.Tensor:
```
Comment (dxoigmn) [inline mart/attack/gain.py:32]: I think `proposals` should be a `Tensor`?

https://github.com/pytorch/vision/blob/0dceac025615a1c2df6ec1675d8f9d7757432a49/torchvision/models/detection/rpn.py#LL243C27-L243C27

File: mart/attack/initializer.py [hunk 9]
```diff
@@ -16,33 +17,33 @@ class Initializer(abc.ABC):
...
         self.constant = constant
 
-    def __call__(self, perturbation):
+    def __call__(self, perturbation: torch.Tensor) -> None:
         torch.nn.init.constant_(perturbation, self.constant)
 
 
 class Uniform(Initializer):
-    def __init__(self, min, max):
+    def __init__(self, min: int, max: int):
```
Comment (dxoigmn) [inline mart/attack/initializer.py:33]: `min` and `max` can be floats.

File: mart/attack/objective/classification.py [hunk 10]
```diff
@@ -13,19 +15,21 @@
...
         # FIXME: I don't like this argmax call. It feels like this should receive input tensor of
         #        the same shape as target?
         mispredictions = input.argmax(dim=-1) != target
         return mispredictions
 
 
 class RandomTarget(Objective):
-    def __init__(self, nb_classes, gain_fn) -> None:
+    def __init__(
+        self, nb_classes: Union[int, float], gain_fn: Callable[torch.Tensor, torch.Tensor]
```
Comment (dxoigmn) [inline mart/attack/objective/classification.py:27]: `nb_classes` should always be an `int`.

File: mart/attack/objective/object_detection.py [hunk 11]
```diff
@@ -18,13 +19,19 @@
 class ZeroAP(Objective):
     """Determine if predictions yields zero Average Precision."""
 
-    def __init__(self, iou_threshold=0.5, confidence_threshold=0.5) -> None:
+    def __init__(
+        self,
+        iou_threshold: Optional[Union[int, float]] = 0.5,
+        confidence_threshold: Optional[Union[int, float]] = 0.5,
```
Comment (dxoigmn) [inline mart/attack/objective/object_detection.py:25]: These will always be `float`s.

File: mart/attack/objective/object_detection.py [hunk 12]
```diff
@@ -60,12 +67,14 @@ class Missed(Objective):
     """The objective of the adversary is to make all AP errors as the missed error, i.e. no object
     is detected, nor false positive."""
 
-    def __init__(self, confidence_threshold=0.5) -> None:
+    def __init__(self, confidence_threshold: Optional[Union[int, float]] = 0.5) -> None:
```
Comment (dxoigmn) [inline mart/attack/objective/object_detection.py:70]: `confidence_threshold` will always be a `float`.

File: mart/attack/perturber/batch.py [hunk 13]
```diff
@@ -19,7 +21,7 @@ class BatchPerturber(Callback, torch.nn.Module):
     We split input into individual examples and run different perturbers accordingly.
     """
 
-    def __init__(self, perturber_factory, *perturber_args, **perturber_kwargs):
+    def __init__(self, perturber_factory: Any, *perturber_args, **perturber_kwargs):
```
Comment (dxoigmn) [inline mart/attack/perturber/batch.py:24]: `perturber_factory` should be `Callable`.

File: mart/attack/projector.py [hunk 14]
```diff
@@ -40,12 +49,19 @@ def __repr__(self):
 class Range(Projector):
     """Clamp the perturbation so that the output is range-constrained."""
 
-    def __init__(self, quantize=False, min=0, max=255):
+    def __init__(
+        self, quantize: Optional[bool] = False, min: Optional[int] = 0, max: Optional[int] = 255
+    ):
```
Comment (dxoigmn) [inline mart/attack/projector.py:54]: `min` and `max` can be `float`s.

File: mart/attack/projector.py [hunk 15]
```diff
@@ -62,12 +78,19 @@ class RangeAdditive(Projector):
     The projector assumes an additive perturbation threat model.
     """
 
-    def __init__(self, quantize=False, min=0, max=255):
+    def __init__(
+        self, quantize: Optional[bool] = False, min: Optional[int] = 0, max: Optional[int] = 255
```
Comment (dxoigmn) [inline mart/attack/projector.py:82]: `min` and `max` can be `float`s.

**Review by dxoigmn — CHANGES_REQUESTED** [review]
See comments.

File: mart/attack/adversary.py (hunk 2 above)
Comment (lumurillo) [reply mart/attack/adversary.py:30]: The `Adversary` class is defined below in this file, so I decided to use its parent class. Other option is to type it as `torch.nn.Module` which is the other parent class.

File: mart/attack/callbacks/base.py (hunk 7 above)
Comment (lumurillo) [reply mart/attack/callbacks/base.py:23]: For the moment I will leave it as `torch.nn.Module` due to a circular import issue when tryin to use the `Adversary` type in this file.

File: mart/attack/adversary_in_art.py (hunk 4 above)
Comment (lumurillo) [reply mart/attack/adversary_in_art.py:27]: Should we use the ART's `ObjectDetector` type in the `target_model` argument?

File: mart/attack/adversary_wrapper.py [hunk 16]
```diff
@@ -16,17 +21,23 @@ class NormalizedAdversaryAdapter(torch.nn.Module):
     External adversaries commonly take input of NCWH-[0,1] and return input_adv in the same format.
     """
 
-    def __init__(self, adversary):
+    def __init__(self, external_adversary: functools.partial):
```
Comment (dxoigmn) [inline mart/attack/adversary_wrapper.py:24]: I still think this should be `Callable` (and not `partial`).

File: mart/attack/perturber/batch.py [hunk 17]
```diff
@@ -19,7 +22,7 @@ class BatchPerturber(Callback, torch.nn.Module):
     We split input into individual examples and run different perturbers accordingly.
     """
 
-    def __init__(self, perturber_factory, *perturber_args, **perturber_kwargs):
+    def __init__(self, perturber_factory: functools.partial, *perturber_args, **perturber_kwargs):
```
Comment (dxoigmn) [inline mart/attack/perturber/batch.py:25]: Should be `Callable` and not `partial`.

**Review by dxoigmn — CHANGES_REQUESTED** [review]
See comments.

File: mart/attack/adversary.py (hunk 2 above)
Comment (lumurillo) [reply mart/attack/adversary.py:29]: Created issue #38 to tackle this. This circular issue can be a sign of a possible design problem between the adversaries and callbacks.

File: mart/attack/adversary_in_art.py (hunk 4 above)
Comment (mzweilin) [reply mart/attack/adversary_in_art.py]: Currently, we expect it to be `ObjectDetector`. But later we may want to use `BaseEstimator` as we generalize the wrapper for non-object detection models in ART.

File: mart/attack/adversary_in_art.py [hunk 18]
```diff
@@ -5,8 +5,12 @@
 # agreement between Intel Corporation and you.
 #
 
+from typing import Any, List, Optional
+
 import hydra
+import numpy
 import torch
+from art.estimators.object_detection.object_detector import ObjectDetector
```
Comment (dxoigmn) [inline mart/attack/adversary_in_art.py:13]: Do you think it's worth guarding against this import using `typing.TYPE_CHECKING`? It seems better to guard imports that are only used when typing checking.

**Review by dxoigmn — COMMENTED** [review]
One comment about using `typing.TYPE_CHECKING` to guard new imports.

File: mart/attack/adversary_in_art.py (hunk 18 above)
Comment (dxoigmn) [reply mart/attack/adversary_in_art.py:13]: I retract my comment :)

**Review by dxoigmn — APPROVED** [review]
LGTM!

**Pull request merged**